packages/fx-core/src/core/middleware/consolidateLocalRemote.ts

```diff
@@ -0,0 +1,315 @@
+  );
+  if (!localEnvExist && !consolidateManifestExist) {
+    return true;
+  }
+  return false;
+}
+
+function outputCancelMessage(ctx: CoreHookContext) {
+  TOOLS?.logProvider.warning(`[core] Upgrade cancelled.`);
+
+  const inputs = ctx.arguments[ctx.arguments.length - 1] as Inputs;
+  if (inputs.platform === Platform.VSCode) {
+    TOOLS?.logProvider.warning(
+      `[core] Notice upgrade to new configuration files is a must-have to continue to use current version Teams Toolkit. If you are not ready to upgrade and want to continue to use the old version Teams Toolkit, please find Teams Toolkit in Extension and install the version <= ????`
+    );
+  } else {
+    TOOLS?.logProvider.warning(
+      `[core] Notice upgrade to new configuration files is a must-have to continue to use current version Teams Toolkit CLI. If you want to upgrade, please trigger this command again.`
+    );
+    TOOLS?.logProvider.warning(
+      `[core] If you are not ready to upgrade and want to continue to use the old version Teams Toolkit CLI, please install the version <= ????`
+    );
+  }
+}
+
+async function consolidateLocalRemote(ctx: CoreHookContext): Promise<boolean> {
+  sendTelemetryEvent(Component.core, TelemetryEvent.ProjectConsolidateUpgradeStart);
+  const inputs = ctx.arguments[ctx.arguments.length - 1] as Inputs;
+  const fileList: Array<string> = [];
+  const loadRes = await loadProjectSettings(inputs, true);
+  if (loadRes.isErr()) {
+    ctx.result = err(loadRes.error);
+    return false;
+  }
+
+  const projectSettings = loadRes.value;
+
+  try {
+    // add local environment
+    const appName = getLocalAppName(projectSettings.appName);
+    sendTelemetryEvent(Component.core, TelemetryEvent.ProjectConsolidateAddLocalEnvStart);
+    const newEnvConfig = environmentManager.newEnvConfigData(appName);
+    const writeEnvResult = await environmentManager.writeEnvConfig(
+      inputs.projectPath!,
+      newEnvConfig,
+      environmentManager.getLocalEnvName()
+    );
+    if (writeEnvResult.isErr()) {
+      throw err(writeEnvResult.error);
+    }
+    fileList.push(path.join(inputs.projectPath as string, ".fx", "configs", "env.local.json"));
+    sendTelemetryEvent(Component.core, TelemetryEvent.ProjectConsolidateAddLocalEnv);
+
+    // add consolidate manifest
+    let manifest: TeamsAppManifest | undefined;
+    const templatesFolder = getTemplatesFolder();
+    const appDir = `${ctx.root}/${APP_PACKAGE_FOLDER_FOR_MULTI_ENV}`;
+    if (isSPFxProject(projectSettings)) {
+      sendTelemetryEvent(Component.core, TelemetryEvent.ProjectConsolidateAddSPFXManifestStart);
+      const templateManifestFolder = path.join(templatesFolder, "plugins", "resource", "spfx");
+      const manifestFile = path.resolve(
+        templateManifestFolder,
+        "./solution/manifest_multi_env.json"
+      );
+      const manifestString = (await fs.readFile(manifestFile)).toString();
+      manifest = JSON.parse(manifestString);
```

Comment:
No need to parse the manifest template either?

Reply:
AppStudio plugin uses the same logic for adding manifest, it prevents manifest format change after packing.

Reply:
Got it, that makes sense.